Applications/Classification/jni/main_func.cpp

```diff
@@ -0,0 +1,305 @@
+  } while (x >= RAND_MAX - remainder);
+  return min + x % n;
+}
+
+/**
+ * @brief     load data at specific position of file
+ * @param[in] F  ifstream (input file)
+ * @param[out] outVec
+ * @param[out] outLabel
+ * @param[in] id th data to get
+ * @retval true/false false : end of data
+ */
+bool getData(std::ifstream &F, std::vector<float> &outVec,
+             std::vector<float> &outLabel, int id) {
+  uint64_t position = (feature_size + total_label_size) * id * sizeof(float);
+  if ((position + (feature_size + total_label_size) * sizeof(float)) >
+      ULLONG_MAX)
+    return false;
+  F.seekg(position, std::ios::beg);
+  for (unsigned int i = 0; i < feature_size; i++)
+    F.read((char *)&outVec[i], sizeof(float));
+  for (unsigned int i = 0; i < total_label_size; i++)
+    F.read((char *)&outLabel[i], sizeof(float));
+
+  return true;
+}
+
+/**
+ * @brief      get data which size is mini batch for train
+ * @param[out] outVec
+ * @param[out] outLabel
+ * @param[out] status for error handling
+ * @retval true/false
+ */
+bool getMiniBatch_train(std::vector<std::vector<std::vector<float>>> &outVec,
+                        std::vector<std::vector<std::vector<float>>> &outLabel,
+                        int &status) {
+  std::vector<int> memI;
+  std::vector<int> memJ;
+  unsigned int count = 0;
+  int data_size = total_train_data_size;
+
+  std::string filename = "trainingSet.dat";
+  std::ifstream F(filename, std::ios::in | std::ios::binary);
+
+  for (unsigned int i = 0; i < total_label_size * data_size; i++) {
+    if (!duplicate[i])
+      count++;
+  }
+
+  if (count < mini_batch) {
+    for (unsigned int i = 0; i < total_label_size * data_size; ++i) {
+      duplicate[i] = false;
+    }
+    return false;
+  }
+
+  count = 0;
+  while (count < mini_batch) {
+    int nomI = rangeRandom(0, total_label_size * data_size - 1);
+    if (!duplicate[nomI]) {
+      memI.push_back(nomI);
+      duplicate[nomI] = true;
+      count++;
+    }
+  }
+
+  for (unsigned int i = 0; i < count; i++) {
+    std::vector<std::vector<float>> out;
+    std::vector<std::vector<float>> outL;
+    std::vector<float> o;
+    std::vector<float> l;
+
+    o.resize(feature_size);
+    l.resize(total_label_size);
+
+    getData(F, o, l, memI[i]);
+
+    out.push_back(o);
+    outL.push_back(l);
+
+    outVec.push_back(out);
+    outLabel.push_back(outL);
+  }
+
+  F.close();
+  return true;
+}
+
+/**
+ * @brief      get data which size is mini batch for train
+ * @param[out] outVec
+ * @param[out] outLabel
+ * @param[out] status for error handling
+ * @retval true/false false : end of data
+ */
+bool getMiniBatch_val(std::vector<std::vector<std::vector<float>>> &outVec,
+                      std::vector<std::vector<std::vector<float>>> &outLabel,
+                      int &status) {
+  std::vector<int> memI;
+  std::vector<int> memJ;
+  unsigned int count = 0;
+  int data_size = total_val_data_size;
+
+  std::string filename = "trainingSet.dat";
```

Comment:
```suggestion
  std::string filename = "valSet.dat";
```

Reply:
It might be better off to allow this function to have `std::string filename` as an argument if putting `trainingSet.dat` is on purpose

Reply:
I don't think so. It is the use case to use function to get the data set. It does not care about how the user provide the data. It could be other way like calculate directly from image or get data from other sources, not from raw file. If someone want to use raw file, then he can specify file name in configuration file. 
And, the reason to use here trainingSet.data instead of valSet.data is that to validate the training mechanism with training set. If we are using training set for validation, the accuracy must be greater than 95% at least. trainignSet.dat is there intentionally.